samples: Add support for different base address of the shared mem in the memory map of each core

iuliana-prodan opened the pull request:
Now, in openamp_rsc_table sample we support only the devices
that have the same value for driver and device physical address.
But, in case they are different, we need to convert from driver
to device address.

This PR introduces address translation support for OpenAMP:
- implements the address translation functions;
- returns the I/O operations used for address translation.

The ops can be used in Libmetal (see metal_io_init()).

Each vendor can add his own addr translation table.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/84168

arnopo commented on samples/subsys/ipc/openamp_rsc_table/src/main_remote.c:
```diff
@@ -36,6 +36,10 @@ LOG_MODULE_REGISTER(openamp_rsc_table, LOG_LEVEL_DBG);
+#define SHM_PHYS_START_ADDR	COND_CODE_1(DT_NUM_REGS(SHM_NODE),  \
+				(DT_REG_ADDR(SHM_NODE)),            \
+				(DT_REG_ADDR_BY_NAME(SHM_NODE, 1))) \
+
```
Something is not clear to me in this patch. It looks like you want to define a virtual address in the device tree, which does not seem like a good idea.

In the device tree, I would expect that the reg property gives only the physical address, which you then map in the driver or application to obtain a virtual address.

But perhaps I am missing something?

iuliana-prodan replied:
@arnopo you understood correctly, in the dts overlay, in `reg` property I'm giving both physical and virtual address.

To do the mapping in the application I'm afraid I might break compatibility for the other boards that have the same physical and virtual address.

Another idea is to take the physical address from a CONFIG which will be defined only if `CONFIG_OPENAMP=y`  and we will configure it only in the board conf that supports address mapping.

@arnopo what do you think?

arnopo replied:
There are two points here.

1)Can we define a virtual address in the device tree? For me, it does not make sense because the hardware reference is the physical address, and the device tree describes the hardware. A virtual address is computed from a physical address, usually with an offset. I don't think there is a precedent for a virtual address defined in the device tree, but perhaps I'm wrong?

2) mapping in the application
  Could you use K_MEM_VIRT_ADDR, which seems to return the physical address if CONFIG_MMU is not enabled?

iuliana-prodan replied:
Both addresses are physical addresses - one from DSP memory space and the other one from main core (Cortex A) memory space - [here ](https://github.com/nxp-imx/linux-imx/blob/lf-6.6.y/arch/arm64/boot/dts/freescale/imx8ulp-evk.dts#L84) is the address in Linux dts for 8ULP.

Virtual address is improper said because DSP doesn't have MMU.
Using K_MEM_VIRT_ADDR wouldn't help because it can't get the correct address, this is nowhere computed in Zephyr.

arnopo replied:
Thanks for the clarification.

What you are trying to do here is convert your Cortex-A memory physical address to the local physical address, which is not related to the virtual address that could be applied on top of it.

In my point of view, you are misusing the OpenAMP API by defining the local physical address as a virtual address.

Memory translation is generally managed by Linux itself using dma-range in the Linux remoteproc. However, this is not implemented for the RPMsg VirtIO buffers, and I suppose that is your issue.

One solution would be to address this on the Linux side. Another would be to implement metal_io_ops to perform the translation if a dma-range is defined in the Zephyr device tree.

I propose we discuss this point in the next RP meeting with the Linux maintainer."

iuliana-prodan replied:
I would choose to implement `metal_io_ops` to do the translation (on Zephyr side), but sure, we can talk to Linux maintainer on RP meeting.

mbolivar reviewed (dismissed):
If I understand you are trying to use ranges to signal virtual and physical address differences.

This is contrary to the DT specification. The ranges property is basically syntactic sugar that makes it easier to write down physical addresses and it shall not be used for virtual addressing.

iuliana-prodan commented:
> If I understand you are trying to use ranges to signal virtual and physical address differences.
> 
> This is contrary to the DT specification. The ranges property is basically syntactic sugar that makes it easier to write down physical addresses and it shall not be used for virtual addressing.

@mbolivar both addresses are physical addresses, as explained [here](https://github.com/zephyrproject-rtos/zephyr/pull/84170#discussion_r1922755716).

[Here's](https://github.com/zephyrproject-rtos/zephyr/pull/83049/commits/906df3d1433d7644b31db8e9fa92fdb1b702e4b4#diff-1c0848c129df6d76185a95f4b292bea15a94de7708891c87c894e677f8de9e8bR18) an example on how I use the `ranges` property for i.MX8ULP where we need to translate from driver physical address (Cortex A) to device physical address (DSP).

If `ranges` is not suitable for this case, do you have any other suggestion?

Thanks!

iuliana-prodan commented:
> I think the solution to this problem would be to use carveouts in the resource table, so the host can store its physical address and the OpenAMP library can translate it to its own device address. This was the reason for my pull request #83754, but I have yet to rewrite the example to finish the pull request.

This solution was agreed in the OpenAMP technical meeting.
See libmetal update here: https://github.com/OpenAMP/libmetal/pull/322

mbolivar commented:
> If `ranges` is not suitable for this case, do you have any other suggestion?

Hi @iuliana-prodan , before we get into the specifics of your use case, can you please confirm that you've read the core DT spec's sections on `reg` and `ranges`, in particular sections 2.3.5 through 2.3.8 in https://devicetree-specification.readthedocs.io/en/latest/index.html ?

I am happy to try to help you with the specifics, but I unfortunately do not have time to educate every zephyr maintainer about the core specification itself. Happy to help if anything in the spec is confusing or if you have further questions after you've confirmed you read the spec.

Thanks!

arnopo commented:
> > If `ranges` is not suitable for this case, do you have any other suggestion?
> 
> Hi @iuliana-prodan , before we get into the specifics of your use case, can you please confirm that you've read the core DT spec's sections on `reg` and `ranges`, in particular sections 2.3.5 through 2.3.8 in https://devicetree-specification.readthedocs.io/en/latest/index.html ?
> 
> I am happy to try to help you with the specifics, but I unfortunately do not have time to educate every zephyr maintainer about the core specification itself. Happy to help if anything in the spec is confusing or if you have further questions after you've confirmed you read the spec.
> 
> Thanks!

In Linux for such translation [dma_ranges](https://devicetree-specification.readthedocs.io/en/latest/chapter2-devicetree-basics.html#dma-ranges) is used.

iuliana-prodan commented:
Changes since v2:
- add `addr_translation` driver that implement `metal_io_ops::phys_to_offset` and `metal_io_ops::offset_to_phys` and return the `metal_io_ops`;
- update `openamp_rsc_table` sample to use the `metal_io_ops` from `addr_translation` driver.

arnopo commented on doc/build/dts/api/api.rst:
```diff
@@ -422,6 +422,10 @@ device.
+   * - zephyr,ipc_shm_translation
+     - A node whose ``reg`` is used by the addr_translation driver to convert,
+       in OpenAMP subsystem, the base address of the shared memory (SHM) from
+       driver to device address, since for some SoCs these are different
```
I don't see any DT overlay that uses this in the openamp_rsc_table. Could point it to me or add it if missing, so that we have a better view of the implementation? 
without more details an regarding this description, it seems to me that you are abusing the reg property usage for the memory translation.

iuliana-prodan replied:
It's here: https://github.com/zephyrproject-rtos/zephyr/pull/83049/files#diff-1c0848c129df6d76185a95f4b292bea15a94de7708891c87c894e677f8de9e8b
I've added a comment [here](https://github.com/zephyrproject-rtos/zephyr/pull/83049#issuecomment-2650708320) - I can put these 2 PRs together.

iuliana-prodan commented:
Changes since v3:
- use `dma-ranges` in addr_translation driver;
- add dts macros for `dma-ranges` property in order to get the needed info.

@arnopo @mbolivar please let me know if you agree with this implementation.

Since in the previous comments you suggested `dma-ranges` for address translation, I had to add support for dts parsing of this property.

iuliana-prodan commented:
Changes since v4:

- add support for i.MX8UPL for `openamp_rsc_table` sample (since https://github.com/zephyrproject-rtos/zephyr/pull/83049 was merged).

arnopo commented on samples/subsys/ipc/openamp_rsc_table/boards/imx8ulp_evk_mimx8ud7_adsp.overlay:
```diff
@@ -0,0 +1,39 @@
+		reg = <0x8fef0000 0x110000>;
+		dma-ranges = <0x1bef0000 0x8fef0000 0x110000>;
+		status = "okay";
+	};
```
Definitely, I think for this PR we need reviews from Zephyr experts to help find the best compromise.

I wonder if this node have a sens in therm of hardware.

Based on Linux experience, I would have in mind a topology based on a bus... Do you think that it would possible to have something like this.

	dsp_bus: @8fef0000 {
		compatible = "nxp,imx8ulp-addr-translation", "simple-bus";  /* define translation in a specific bus */
		#address-cells = <1>;
		#size-cells = <1>;   
		ranges;    /* This is mandatory in  DT schema  to enable dma-ranges*/
		dma-ranges = <0x1bef0000 0x8fef0000 0x110000>;  /* translation between the device to loacal address */

		dspsram3: memory@0x8fef0000 {   
			reg = <0x8fef0000 0x110000>;   /* local address */ 
		};

		status = "okay";
	};
	
Then your "zephyr,addr-translation" would become a bus with ops for the translation

iuliana-prodan replied:
> Definitely, I think for this PR we need reviews from Zephyr experts to help find the best compromise.

@carlescufi @nashif Could you please add reviewers who might assist with this PR? 

To summarize, this PR includes an address translation driver currently utilized in the `openamp_rsc_table` sample. This driver translates addresses from the driver's perspective (as seen by Linux) to the device's perspective (as seen by Zephyr).

doki-nordic reviewed (dismissed):
I am not devicetree specialist, so I didn't get too deep into DTS details. From IPC point of view, look good to me.

iuliana-prodan commented:
> I am not devicetree specialist, so I didn't get too deep into DTS details. From IPC point of view, look good to me.

Thanks for reviewing @doki-nordic.
I was waiting for a feedback from @mbolivar to see if it's ok to add macros for `dma-ranges`.
If the community agrees with this, I'll need to add comments, some examples, etc.

iuliana-prodan commented on include/zephyr/devicetree.h:
```diff
@@ -2189,6 +2189,39 @@
+ * @{
+ */
+
+#define DT_NUM_DMA_RANGES(node_id) DT_CAT(node_id, _DMA_RANGES_NUM)
```
As I already mentioned, if the community agrees with this, I'll need to add comments, examples, some tests, etc.

iuliana-prodan committed:
openamp: add core address translation support

There are SoCs that have different memory maps for device and driver.
Therefore the shared memory between them has different values and we
need to convert the addresses.

This commit introduces address translation support for OpenAMP:
- implements the address translation functions;
- returns the I/O operations used for address translation.
The ops can be used in Libmetal (see metal_io_init()).

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
openamp: add NXP address translation table

Add NXP-specific address translation table to support memory-mapped
translations between driver and device.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan commented:
Changes since `v4`:

Since no feedback received for `dma-ranges` in addr_translation driver, I've reworked the PR:
- add core address translation functionality to OpenAMP to support SoCs with differing memory maps between devices and drivers;
- each vendor can add his own address translation table - added an example for NXP;
- update openamp_rsc_table sample to use `metal_io_get_ops()` for address translation, if configured;
- enable address translation for i.MX8ULP.

arnopo reviewed (commented):
This approach seems much better to me. I have suggested some improvements, hoping that I did not miss anything important in my proposal.

iuliana-prodan committed:
openamp: add Kconfig entries for address translation

Add configuration options to enable address translation in OpenAMP
and specify vendor header files for translation tables.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
openamp: enable address translation in CMake

Update CMakeLists.txt to conditionally include address translation
support based on Kconfig options.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
samples: enable OpenAMP address translation on i.MX8ULP

Enable address translation in board config for imx8ulp_evk_mimx8ud7_adsp.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
samples: use address translation ops in openamp_rsc_table sample

Update openamp_rsc_table sample to use `metal_io_ops` for address
translation, if configured.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan commented on lib/open-amp/addr_translation/addr_translation.h:
```diff
@@ -0,0 +1,15 @@
+ * @param	dev	Driver instance pointer
+ * @return	metal_io_ops struct
+ */
+const struct metal_io_ops *metal_io_get_ops(void);
```
done something similar. I haven't used` #ifdef CONFIG_OPENAMP_ADDR_TRANSLATION` in the sample.

iuliana-prodan commented on lib/open-amp/addr_translation/addr_translation.c:
```diff
@@ -0,0 +1,143 @@
+struct tlb_entries {
+	const metal_phys_addr_t *dev_addr;	/* device address */
+	const metal_phys_addr_t *drv_addr;	/* driver address */
+};
```
I've used a similar structure and another one that includes the `no_pages` so no `CONFIG_OPENAMP_NB_TRANSLATION_REGIONS` is needed.

iuliana-prodan commented on lib/open-amp/addr_translation/addr_translation.c:
```diff
@@ -0,0 +1,143 @@
+/*
```
I've removed the `addr_translation` folder, the file is added near` resource_table.c`

rruuaanng commented on lib/open-amp/addr_translation.c:
```diff
@@ -0,0 +1,125 @@
+ * In this file addr_map needs to be defined,
+ * see example in nxp_addr_translation.h
+ */
+#ifdef CONFIG_OPENAMP_VENDOR_ADDR_TRANSLATION
+#include CONFIG_OPENAMP_VENDOR_ADDR_TRANSLATION_FILE
+struct regions *addr_map = &vendor_addr_map;
```
Is this style allowed by us?
(Just a question)

iuliana-prodan commented:
Changes since `v6`:
- rebased and resolved conflicts with PR #87296
- replaced hardcoded variable name with a getter function

iuliana-prodan commented:
> @iuliana-prodan This needs a rebase.

done

arnopo commented on lib/open-amp/addr_translation.c:
```diff
@@ -0,0 +1,131 @@
+		if ((tmp_offset <= offset) && (offset < tmp_offset + tmp_size)) {
+			return tmp_phys + (offset - tmp_offset);
+		}
+		tmp_offset += tmp_size;
```
The offset is  related to the bas address of the io_region. Seems to me not good, but perhaps I'm wrong?
I would expect to see here the use of  `metal_io_phys()` to get the physical address and then check that this address is in the memory translation region.

iuliana-prodan replied:
I've pushed a new version.

I renamed and clarified the structures used.
> you have only one io regions
> the addr_map->regions[0].addr must matche with to the base address of your io region

I added the selection of the table of base physical addresses for the I/O region, based on the given physical base address - see `get_phys_map(metal_phys_addr_t phys)`

> the addr_map->regions[] should be continuous.

The physical pages must be mapped continuously in the virtual memory space. If the virtual and physical addresses are the same, then, of course, the pages must be continuous. But, I don't see why I should enforce anything here.

> All this could be avoid if you test io->physmap[0] + offset or 'io->virt + offset'

I validated the offset with the io->size.

arnopo replied:
Ok, I still not convince that this code is robust but it answer for your need and can be fixed later if needed.

iuliana-prodan commented on lib/open-amp/addr_translation.c:
```diff
@@ -0,0 +1,131 @@
+
+	if (!addr_map) {
+		return METAL_BAD_PHYS;
+	}
```
removed

iuliana-prodan commented:
Changes since `v7`:
- renamed the structures used for address translation, based on @arnopo's suggestion;
- removed unnecessary null pointer checks.

tejlmand commented on lib/open-amp/vendor/nxp_addr_translation.h:
```diff
@@ -0,0 +1,23 @@
+
+static const struct addr_trans_region regions[] = {
+	{
+		.addr = 0x1bef0000,
+		.remote_addr = 0x8fef0000,
+		.size = 0x10000000
```
minor nit. Would be nice with comment informing from where those values originates.

iuliana-prodan replied:
added comments -see https://github.com/zephyrproject-rtos/zephyr/pull/84170/files#diff-8b96f0f54f5cf721c371626c99fde50de1f3fdf0ca53445435d33144796c52c7R10

iuliana-prodan commented:
Changes since `v8`:
- renamed structures and added additional comments for improved clarity and readability;
- implemented selection logic for the base physical address table of the I/O region, based on the provided physical base address - see `addr_translation_get_ops(metal_phys_addr_t phys)` [here](https://github.com/zephyrproject-rtos/zephyr/pull/84170/files#diff-df98bcf9fa890735f36f42c52d2dbe952b97455500e60ea1a6184e66e3826356R126);
- validated the `offset` against `io->size` to ensure correctness.
